book/src/designs/0002-parallel-verification.md

```diff
@@ -0,0 +1,426 @@
+
+This RFC focuses on the verification stages that use chain context, and the
+design patterns that enable them to be excuted in parallel.
+
+## Verification Interfaces
+[verification-interfaces]: #verification-interfaces
+
+Verifcation is implemented by the following traits and services:
+* **Structural Verification:**
+  * `zebra_chain::ZcashDeserialize`: A trait for parsing consensus-critical
+    data structures from a byte buffer.
+  * `zebra_network::init`: Provides a downloader service that accepts a
+    `BlockHeaderHash` request, and parses the peer response into a `Block`
+    struct.
+* **Context-Free Verification:**
+  * `zebra_consensus::ContextFreeVerifier`: Provides a verifier service that
+    accepts a `Block` request, performs context-free verification on the block,
+    and responds with an identifier for the block.
+  * Note: as of 27 July 2020, context-free verification is partly implemented
+    by `BlockVerifier`, which responds with `BlockHeaderHash`.
+* **Prospective Verification:**
+  * `zebra_consensus::ProspectiveVerifier`: Provides a verifier service that
+    accepts a `Block` request, performs prospective verification on the block,
+    and responds with the block's `VerificationConstraints`, which include a
+    reference to the `Block` itself.
+* **Constraint Verification:**
+  * `zebra_consensus::ConstraintVerifier`: Provides a verifier service that
+    accepts a `VerificationConstraints, previous_context: ChainContext`
+    request, performs constraint checks on the `Block` and `ChainContext`, and
+    responds with an idenfier for the block.
+  * Note: blocks can be identified by their BlockHeaderHash, or by reference.
+* **Contextual Verification:**
+  * `zebra_consensus::ContextualVerifier`: Provides a verifier service that
+    accepts a `Block, previous_context: ChainContext` request, performs any
+    remaining contextual verification on the block, and responds with an
+    identifier for the block.
+* **In-Memory Chain Updates:**
+  * `zebra_consensus::RecentChainUpdater`: Provides an updater service that
+    accepts a `Block, previous_context: ChainContext` request, generates an
+    updated chain context, dropping references to any ancestor blocks that are
+    no longer required to verify subsequent blocks. The service responds with
+    an updated `ChainContext`, which contains a reference to the block itself.
+* **Chain Tip Updates:**
+  * `zebra_consensus::ChainTipUpdater`: Provides an updater service that
+    accepts a `ChainContext, Arc<Mutex<ChainTips>>>` request, updates
+    the set of chain tips, and updates the main chain tip, if required.
+    (New blocks on side-chains might not lead to a main chain tip update.)
+    The `ChainContext` becomes one of the new tips in the list.
+    The service responds with the updated `Arc<Mutex<ChainTips>>`.
+  * Note: The service requires exclusive write access to the chain tips, so it
+    can atomically update the main tip and the set of chain tips.
```

Comment:
> Not sure about this, the usage of an arc mutex makes me think that we're hacking around the real data flow that is happening here. I'm assuming this ChainTips type is owned by the state service and we're passing a mutable reference to it out via the arc mutex.

I think you're right, the state service should own the tips, and provide access by returning a read-only set. (Or the main tip.)

> I thought the plan was to have the ChainContext be a trait where the type is defined by the `zebra-consensus` crate with an `Ord` implementation that is utilized by the state service to sort them and pick the main chain according to `zebra-consensus`'s own rules.

I don't think we can implement use Ord to implement the "avoid replacing an existing main tip, but break empty tip ties using the block hash" rule:
https://github.com/ZcashFoundation/zebra/pull/763/files#diff-5d14d1a9af09715ab941449d29b4a605R319

So we might need to pass a custom trait or function instead.

Reply:
> I don't think we can implement use Ord to implement the "avoid replacing an existing main tip, but break empty tip ties using the block hash" rule:

Is the only concern the rule where if we get a tie and one of them is already the main chain we use that one? As far as I can tell everything else can easily live in the `Ord` impl. As for the sort stability concern, if its already not a consensus rule I think its okay to let that be handled by the state. Worst case scenario we could make the sorting function be part of the `ChainContext` trait that we expose in `zebra-state` so the logic still lives in `zebra-consensus` and is just called in `zebra-state`.

Reply:
To be clear, it's not a consensus rule, but it's important, because it helps the network converge on a single chain.

That said, if the state requires "strictly greater than" work before changing the tip, online network convergence should work.

I think verification requires that all blocks are part of a chain, so might be able to rule out final edge case. If we can't, the worst-case scenario is that we pick the first equal block from disk or memory, which is unlikely to cause any problems in practice. (If the whole network crashes and restarts at the same time, we have other issues.)

Reply:
So, based on recent discussions about who is in charge of which stages of the verification I'm assuming that the sorting rules are the responsibility of the `zebra-state` crate.

Reply:
Moved to #871, I split off the state-specific sections into their own PR.